plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java

```diff
@@ -39,14 +46,22 @@ protected DeployPreferencesDialog newDeployPreferencesDialog(Shell shell, IProje
   }
 
   @Override
-  protected StagingDelegate getStagingDelegate(IProject project) {
+  protected StagingDelegate getStagingDelegate(IProject project) throws CoreException {
     String appYamlPath = new FlexDeployPreferences(project).getAppYamlPath();
     IFile appYaml = project.getFile(appYamlPath);
     if (!appYaml.exists()) {
       throw new IllegalStateException("Invalid app.yaml path from project preferences.");
     }
     IPath appEngineDirectory = appYaml.getParent().getLocation();
-    return new FlexStagingDelegate(appEngineDirectory);
+
+    IFacetedProject facetedProject = ProjectFacetsManager.create(project);
+    if (AppEngineFlexWarFacet.hasFacet(facetedProject)) {
+      return new FlexStagingDelegate(appEngineDirectory);
+    }
+    if (AppEngineFlexJarFacet.hasFacet(facetedProject) && MavenUtils.hasMavenNature(project)) {
+      return new FlexJarMavenProjectStagingDelegate(appEngineDirectory);
```

Comment:
What if it has the Jar facet but is not a Maven project?

Reply:
It cannot happen because the deploy command is not enabled in that case (for now), but if it can ever happen, it will throw the runtime exception below.